config/configtls/configtls.go

```diff
@@ -201,22 +202,97 @@ func (c TLSSetting) loadTLSConfig() (*tls.Config, error) {
-	caPEM, err := os.ReadFile(filepath.Clean(caPath))
+func (c TLSSetting) loadCertPool() (*x509.CertPool, error) {
+	// There is no need to load the System Certs for RootCAs because
+	// if the value is nil, it will default to checking against th System Certs.
+	var err error
+	var certPool *x509.CertPool
+
+	switch {
+	case len(c.CAFile) != 0 && len(c.CAPem) != 0:
+		return nil, fmt.Errorf("failed to load CA CertPool: CA File and PEM cannot both be provided")
+	case len(c.CAFile) != 0:
+		// Set up user specified truststore from file
+		certPool, err = c.loadCertFile(c.CAFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA CertPool File: %w", err)
+		}
+	case len(c.CAPem) != 0:
+		// Set up user specified truststore from PEM
+		certPool, err = c.loadCertPem(c.CAPem)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA CertPool PEM: %w", err)
+		}
+	}
+
+	return certPool, nil
+}
+
+func (c TLSSetting) loadCertFile(caPath string) (*x509.CertPool, error) {
+	caPEM, err := ioutil.ReadFile(filepath.Clean(caPath))
 	if err != nil {
 		return nil, fmt.Errorf("failed to load CA %s: %w", caPath, err)
 	}
 
+	return c.loadCertPem(caPEM)
+}
+
+func (c TLSSetting) loadCertPem(caPEM []byte) (*x509.CertPool, error) {
 	certPool := x509.NewCertPool()
 	if !certPool.AppendCertsFromPEM(caPEM) {
-		return nil, fmt.Errorf("failed to parse CA %s", caPath)
+		return nil, fmt.Errorf("failed to parse CA %s", caPEM)
 	}
 	return certPool, nil
 }
 
+func (c TLSSetting) loadCertificate() (tls.Certificate, error) {
+	hasCertFile := len(c.CertFile) != 0
+	hasCertPem := len(c.CertPem) != 0
+	hasKeyFile := len(c.KeyFile) != 0
+	hasKeyPem := len(c.KeyPem) != 0
+	hasCert := hasCertFile || hasCertPem
+	hasKey := hasKeyFile || hasKeyPem
+
+	// Validate the incoming parameters.
+	switch {
+	case hasCert != hasKey:
+		return tls.Certificate{}, fmt.Errorf("for auth via TLS, either both certificate and key must be supplied, or neither")
+	case hasCertFile && !hasKeyFile:
+		return tls.Certificate{}, fmt.Errorf("failed to load TLS cert file and key PEM: both must be provided as a file or both as a PEM")
+	case !hasCertFile && hasKeyFile:
+		return tls.Certificate{}, fmt.Errorf("failed to load TLS cert PEM and key file: both must be provided as a file or both as a PEM")
+	case hasCertFile && hasCertPem:
+		return tls.Certificate{}, fmt.Errorf("for auth via TLS, certificate file and PEM cannot both be provided")
+	case hasKeyFile && hasKeyPem:
+		return tls.Certificate{}, fmt.Errorf("for auth via TLS, key file and PEM cannot both be provided")
+	}
+
+	// If we don't have a cert, return the nil pointer.
+	if !hasCert {
+		return tls.Certificate{}, nil
+	}
+
+	// Add the tls cert to the certificates from the files or in memory PEM.
+	var err error
+	var certificate tls.Certificate
+	if hasCertFile {
+		certificate, err = tls.LoadX509KeyPair(filepath.Clean(c.CertFile), filepath.Clean(c.KeyFile))
```

Comment:
`tls.LoadX509KeyPair` just reads the files from disk (os.ReadFile) and calls `tls.X509KeyPair`, I think we can probably drop the first call and have our own logic for getting the pem content either from the file or the inline strings.